File: src/main/kotlin/codeacademy/bookingforum/app/topicCategory/TopicCategoryService.java
package codeacademy.bookingforum.app.topicCategory;

import codeacademy.bookingforum.app.configuration.ResponseObject;
import codeacademy.bookingforum.app.user.auth.UserAuth;
import codeacademy.bookingforum.app.user.auth.UserAuthRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.WebRequest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Service
public class TopicCategoryService {
    @Autowired
    TopicCategoryRepo sectionRepo;
    @Autowired
    TopicCategoryMapper sectionMapper;
    @Autowired
    UserAuthRepo userRepo;

    // Create new section
    public ResponseObject create(TopicCategoryDto section, WebRequest request) {
        sectionRepo.save(sectionMapper.fromDto(section));
        return new ResponseObject(Collections.singletonList("Section created successfully."), HttpStatus.CREATED, request);
    }

    // Get a list of all sections, filter through depending on authorization state, return modified list of sections
    public List<TopicCategoryDto> getList() {
        List<TopicCategory> sections = (List<TopicCategory>) sectionRepo.findAll();
        UserDetails userDetails = null;

        try {userDetails = (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        } catch (Exception e) {assert true;}

        if (userDetails != null) {
            UserAuth user = userRepo.findByUsername(userDetails.getUsername());
            List<String> userRoles = new ArrayList<>();
            user.getRoles().forEach(role -> userRoles.add(role.getDisplayName()));

            sections.forEach(section -> section.getRoles().forEach(role -> {
                if (!userRoles.contains(role)) {
                    sections.remove(section);
                }
            }));
        } else {
            sections.forEach(section -> {
                if (section.getRoles().contains("ROLE_ADMIN") || section.getRoles().contains("ROLE_SELLER")) {
                    sections.remove(section);
                }
            });
        }

        return sectionMapper.toDtoList(sections);
    }
}

File: src/main/kotlin/codeacademy/bookingforum/app/topicCategory/TopicCategoryService.kt
package codeacademy.bookingforum.app.topicCategory

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class TopicCategoryService {
    @Autowired
    private lateinit var mapper: TopicCategoryMapper

    @Autowired
    private lateinit var repository: TopicCategoryRepository

    fun getAllTopicCategories(): List<TopicCategoryDto?> {
        return repository.findAll().map { mapper.toDto(it) }
    }

    fun getAllTopicCategoryById(id: Long): TopicCategoryDto? {
        return repository.findByIdOrNull(id).let { mapper.toDto(it) }
    }

    fun postTopicCategory(topicCategoryDto: TopicCategoryDto?) {
        mapper.fromDto(topicCategoryDto)?.let { repository.save(it) }
    }

    fun updateTopicCategory(id: Long, topicCategoryDto: TopicCategoryDto?): String {
        return if (repository.existsById(id)) {
            TopicCategory(
                id,
                topicCategoryDto?.title,
                topicCategoryDto?.description,
                topicCategoryDto?.roles
            ).let { repository.save(it) }
            "Topic category with id $id updated"
        } else {
            "Topic category with id $id not found"
        }
    }

    fun deleteTopicCategory(id: Long): String {
        return if (repository.existsById(id)) {
            repository.deleteById(id)
            "Deleted user with id $id"
        } else {
            "Id not found"
        }
    }
}